---
title: "Untitled"
author: "Andrew Dominguez"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Load Fastas}
library(Biostrings)

#imports fasta reference file
fastaImport <- function(fasta){
  fastaFile <- readDNAStringSet(fasta)
  seq_name = names(fastaFile)
  sequence = paste(fastaFile)
  Transcript.Dataframe <- data.frame(seq_name, sequence)
  return(Transcript.Dataframe)
}

```

```{r}
library(ggplot2)
Cbp80 <- fastaImport("Outputs/Cbp80/FASTA/Cbp80_cDNA.fa")
Cbp80.length <- vector("numeric", length(Cbp80$sequence))
for (i in seq_along(Cbp80$sequence)){
  Cbp80.length[i] <- nchar(Cbp80$sequence[i]) 
}

Cbp80Hpr1 <- fastaImport("Outputs/Cbp80Hpr1/FASTA/Cbp80Hpr1_cDNA.fa")
Cbp80Hpr1.length <- vector("numeric", length(Cbp80Hpr1$sequence))
for (i in seq_along(Cbp80Hpr1$sequence)){
  Cbp80Hpr1.length[i] <- nchar(Cbp80Hpr1$sequence[i]) 
}

Cbp80.length.df <- as.data.frame(Cbp80.length)
Cbp80.length.df$Condition <- "Cbp80"
colnames(Cbp80.length.df)[1] <- "Length"

Cbp80Hpr1.length.df <- as.data.frame(Cbp80Hpr1.length)
Cbp80Hpr1.length.df$Condition <- "Cbp80Hpr1"
colnames(Cbp80Hpr1.length.df)[1] <- "Length"
Length.df <- rbind(Cbp80.length.df, Cbp80Hpr1.length.df)

ggplot(Length.df, aes(x=Length, fill=Condition)) +
  geom_density(alpha=0.4) +
  geom_vline(xintercept = 200, colour="black", linetype = "longdash") +
  xlim(c(0,10000))

```

Plan to use RNApar to generate ifragments for longer transcripts and then 
predict structure using Ufold

Fragments less than 200bps need to be filled with "-"

```{r}
library(tidyverse)
new.colnames <- c("seq_name", "Group")
Cbp80.Comms <- read.csv("Outputs/Cbp80/SEEKR/Cbp80_comms.csv")
Cbp80.Comms$X <- substring(Cbp80.Comms$X, 2)
colnames(Cbp80.Comms) <- new.colnames
Cbp80Hpr1.Comms <- read.csv("Outputs/Cbp80Hpr1/SEEKR/Cbp80Hpr1_comms.csv")
Cbp80Hpr1.Comms$X <-substring(Cbp80Hpr1.Comms$X, 2)
colnames(Cbp80Hpr1.Comms) <- new.colnames

Cbp80 <- merge(Cbp80,Cbp80.Comms, by= "seq_name")
Cbp80Hpr1 <- merge(Cbp80Hpr1,Cbp80Hpr1.Comms, by= "seq_name")

```

***if both "mins" are 200, the samples have been "padded" and are ready to be input into RNApar***
Samples will need to be converted to fasta with their RNA sequences

```{r}
library(clusterProfiler)
library(org.Sc.sgd.db)

#function taken from https://bootstrappers.umassmed.edu/guides/main/r_writeFasta.html
writeFasta<-function(data, filename){
  fastaLines = c()
  for (rowNum in 1:nrow(data)){
    fastaLines = c(fastaLines, as.character(paste(">", data[rowNum,"seq_name"], sep = "")))
    fastaLines = c(fastaLines,as.character(data[rowNum,"sequence"]))
  }
  fileConn<-file(filename)
  writeLines(fastaLines, fileConn)
  close(fileConn)
}

GeneSubset<-function(data, Community){
  plotname <- deparse(substitute(data))
  GeneSet <- subset(data, data$Group==Community)
  GeneSet.GO <- enrichGO(gene   = GeneSet$seq_name,
                OrgDb         = org.Sc.sgd.db,
                keyType       = 'ORF',
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05)
  GO.plot <- dotplot(GeneSet.GO, showCategory=10) + ggtitle(paste0(plotname,"_", Community))
  return(list(GeneSet, plot(GO.plot)))
}
#specify which subsets to generate Fasta files for
```

```{r}
library(patchwork)
Cbp80_0 <- GeneSubset(Cbp80, 0)
Cbp80_1 <- GeneSubset(Cbp80, 1)
Cbp80_2 <- GeneSubset(Cbp80, 2)
Cbp80_3 <- GeneSubset(Cbp80, 3)
Cbp80_4 <- GeneSubset(Cbp80, 4)
Cbp80_5 <- GeneSubset(Cbp80, 5)

Cbp80Hpr1_0 <- GeneSubset(Cbp80Hpr1, 0)
#Cbp80Hpr1_1 <- GeneSubset(Cbp80Hpr1, 1) No GO terms for this gene set
Cbp80Hpr1_2 <- GeneSubset(Cbp80Hpr1, 2)
Cbp80Hpr1_3 <- GeneSubset(Cbp80Hpr1, 3)
Cbp80Hpr1_4 <- GeneSubset(Cbp80Hpr1, 4)
Cbp80Hpr1_5 <- GeneSubset(Cbp80Hpr1, 5)
```

Input these newly formatted fasta files into RNApar

```{r Generate Modifications motifs}
ModificationMotifs <- read.csv("RMBase_sacCer3_all_MODs_site.csv")
CDSMotifs <- subset(ModificationMotifs, ModificationMotifs$GeneType == "protein_coding")

RNAprep <- function(sequence) {
  RNA.seq <- vector("numeric", length(sequence))
  for (i in seq_along(sequence)){
    RNA.seq[i] <- gsub("T", "U", sequence[i])
  }
  #seq.fill <- str_pad(RNA.seq, 200, pad = "-", side = "right")
  return(RNA.seq)
}

writeFasta<-function(data, filename){
  fastaLines = c()
  for (rowNum in 1:nrow(data)){
    fastaLines = c(fastaLines, as.character(paste(">", data[rowNum,"Mod_ID"], sep = "")))
    fastaLines = c(fastaLines,as.character(data[rowNum,"RNA_Sequence"]))
  }
  fileConn<-file(filename)
  writeLines(fastaLines, fileConn)
  close(fileConn)
}

CDSMotifs$RNA_Sequence <- RNAprep(CDSMotifs$Sequence)
writeFasta(CDSMotifs, "RNA_modifications.fa")
```


