---
title: "Yra1 RNA-seq Transcript Characterization analysis"
output:
  pdf_document: default
  html_document: default
classoption: landscape
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, include = T, eval = T, warning = F,
                      cache = F)
```

```{r libraries}
library(dplyr)
library(readr)
library(tibble)
library(tidyr)
library(ggplot2)
library(DESeq2)
library(UpSetR)
library(xlsx)
library(rtracklayer)
library(ggpmisc)
library(ggpubr)
library(broom)
library(kableExtra)
library(clusterProfiler)
```

```{r helper_functions}
orf_to_common <- function(keys){
  commons <- AnnotationDbi::select(org.Sc.sgd.db::org.Sc.sgd.db,
                                   keys = keys,
                                   columns=c("COMMON"),
                                   keytype="ORF")
  commons <- commons[match(unique(commons$ORF), commons$ORF), ]
  commons <- commons[ , -2]
  commons$COMMON <- ifelse(is.na(commons$COMMON), commons$ORF, commons$COMMON)
  commons <- unique(commons[ , ])
  commons <- commons$COMMON
  return(commons)
}

orf_to_description <- function(keys){
  description <- AnnotationDbi::select(org.Sc.sgd.db::org.Sc.sgd.db,
                                       keys = keys,
                                       columns=c("DESCRIPTION"),
                                       keytype="ORF")
  description <- description[match(unique(description$ORF), description$ORF), ]
  description <- description$DESCRIPTION
  return(description)
}
```

```{r Loading data and conditions}
logfc_threshold = 1

info <- read_csv("inputs/info.csv")
info$condition<- factor(info$condition, levels = c("Yra1-Input", 
                                                   "Yra1-IP",
                                                   "Cbp80-Input",
                                                   "Cbp80-IP",
                                                   "Hpr1-Input",
                                                   "Hpr1-IP",
                                                   "Tho2Del-Input", 
                                                   "Tho2Del-IP", 
                                                   "Tho2DelCbp80-Input", 
                                                   "Tho2DelCbp80-IP")) #The conditions here need to be changed for Ryutas experiment (Input, HPR3 IP...etc)
counts <- read_csv("inputs/raw_counts.csv") %>%
  #filter(!grepl("__", gene)) %>%
  #select(!starts_with("AP_D")) %>%
  column_to_rownames("gene")
counts <- counts[,-1]
```

```{r deseq2 and function setup}
dds <- DESeqDataSetFromMatrix(counts,
                              colData = info,
                              design = ~ condition)
ds <- DESeq(dds, test="Wald")
#resultsNames(ds)
# log2 fold change (MLE): condition total vs input
# results are log2(total/input)
# log2FC values are expressed as that change in gene expression in total compared
# to that in input.


DE_Compare <- function(comparator, baseline) {
  resultsDF <- results(ds, contrast = c("condition", comparator, baseline), alpha = .05) #here comparator should be IP and Baseline should be input
  resultsDF <- resultsDF %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05) %>%
  filter(abs(log2FoldChange) > logfc_threshold) %>%
  mutate(common = orf_to_common(gene)) %>%
  mutate(description = orf_to_description(gene)) %>%
  na.omit()
  resultsDF <- resultsDF[resultsDF$gene!=resultsDF$common,]
  
}

```


```{r diffex for each strain}
#Comparison between strain inputs
res_Cbp80vNoTag_InputBackground <- DE_Compare("Cbp80-Input", "Yra1-Input")
res_Hpr1vNoTag_InputBackground <- DE_Compare("Hpr1-Input", "Yra1-Input")
res_Tho2Cbp80vTho2_InputBackground <- DE_Compare("Tho2DelCbp80-Input", "Tho2Del-Input")

#Comparison between strain IPs
res_Cbp80vNoTag <- DE_Compare("Cbp80-IP", "Yra1-IP")
res_Hpr1vNoTag <- DE_Compare("Hpr1-IP", "Yra1-IP")
res_Tho2Cbp80vTho2 <- DE_Compare("Tho2DelCbp80-IP", "Tho2Del-IP")

# write.xlsx(res_total, file="outputs/xlsx/deseq2_results_background.xlsx", sheetName="res_total", row.names=FALSE)
```

```{r remove_specific_background}
# removes background the genes that are differentially expressed between Input and IP conditions
remove_background <- function(res_IP, res_Input){
  res_Input_up <- res_Input %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(padj < 0.05)
  
  res_Input_down <- res_Input %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(padj < 0.05)
  
  background_up <- c(res_Input_up$gene)
  background_down <- c(res_Input_down$gene)
  
  res_up <- res_IP %>%
    filter(log2FoldChange > logfc_threshold) %>%
    filter(!gene %in% background_up)
  
  res_down <- res_IP %>%
    filter(log2FoldChange < -logfc_threshold) %>%
    filter(!gene %in% background_down)
  
  res_no_background <- rbind(res_up, res_down)
  return(res_no_background)
}




res_Cbp80 <- remove_background(res_Cbp80vNoTag, res_Cbp80vNoTag_InputBackground)
res_Hpr1 <- remove_background(res_Hpr1vNoTag, res_Hpr1vNoTag_InputBackground)
res_Tho2Cbp80 <- remove_background(res_Tho2Cbp80vTho2, res_Tho2Cbp80vTho2_InputBackground)

# write.xlsx(res_total_ngbp, file="outputs/xlsx/deseq2_results_no_specific_background.xlsx", sheetName="res_total_ngbp", row.names=FALSE)
```



### This report was run with a log2FC threshold of `r logfc_threshold`



## Genes enriched and depleted in Yra1 containing IPs   

```{r upset_plot, fig.height=8, fig.width = 12}
# upset plot after removal (nsb - no specific background) ----------------------
up_nsb <- fromList(list("Cbp80" = filter(res_Cbp80, log2FoldChange > logfc_threshold & padj <0.05)$gene,
                        'Hpr1' = filter(res_Hpr1, log2FoldChange > logfc_threshold & padj <0.05)$gene))

upset_up <- upset(up_nsb, nsets = 2, set_size.show = T, set_size.scale_max = 1800)
upset_up_cp<- cowplot::plot_grid(NULL, 
                                 upset_up$Main_bar, upset_up$Sizes, upset_up$Matrix,
                                 nrow=2, align='hv', rel_heights = c(2,1),
                                 rel_widths = c(2,2))

down_nsb <-  fromList(list("Cbp80" = filter(res_Cbp80, log2FoldChange < -logfc_threshold & padj <0.05)$gene,
                           'Hpr1' = filter(res_Hpr1, log2FoldChange < -logfc_threshold & padj <0.05)$gene))
upset_down <- upset(down_nsb, nsets = 2, set_size.show = T, set_size.scale_max = 1800)
upset_down_cp <- cowplot::plot_grid(NULL, 
                                    upset_down$Main_bar, upset_down$Sizes, upset_down$Matrix,
                                    nrow=2, align='v', rel_heights = c(2,1),
                                    rel_widths = c(2, 2))

gridExtra::grid.arrange(upset_up_cp, upset_down_cp, ncol = 2,
                        top = grid::textGrob("Enriched                                                                 Depleted"))
```

\newpage



```{r separate_shared_sets}
# separate genes from 7 permutations, no specific background -------------------
fromList2 <- function (input) {
  # Same as original fromList()...
  elements <- unique(unlist(input))
  data <- unlist(lapply(input, function(x) {
    x <- as.vector(match(elements, x))
  }))
  data[is.na(data)] <- as.integer(0)
  data[data != 0] <- as.integer(1)
  data <- data.frame(matrix(data, ncol = length(input), byrow = F))
  data <- data[which(rowSums(data) != 0), ]
  names(data) <- names(input)
  # ... Except now it conserves your original value names!
  row.names(data) <- elements
  return(data)
}

get_intersect_members <- function (x, ...){
  require(dplyr)
  require(tibble)
  x <- x[,sapply(x, is.numeric)][,0<=colMeans(x[,sapply(x, is.numeric)],na.rm=T) & colMeans(x[,sapply(x, is.numeric)],na.rm=T)<=1]
  n <- names(x)
  x %>% rownames_to_column() -> x
  l <- c(...)
  a <- intersect(names(x), l)
  ar <- vector('list',length(n)+1)
  ar[[1]] <- x
  i=2
  for (item in n) {
    if (item %in% a){
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '>= 1')
        i <- i + 1
      }
    } else {
      if (class(x[[item]])=='integer'){
        ar[[i]] <- paste(item, '== 0')
        i <- i + 1
      }
    }
  }
  do.call(filter_, ar) %>% column_to_rownames() -> x
  return(x)
}

format_intersect_members <- function(intersect_members_df){
  intersect_members_df <- intersect_members_df %>%
    rownames_to_column("gene") %>%
    mutate(common = orf_to_common(gene)) %>%
    mutate(description = orf_to_description(gene))
  return(intersect_members_df)
}

# res enriched in Cbp80
# res enriched in Cbp80 & Hpr1


up_nsb2 <- fromList2(list("Cbp80" = filter(res_Cbp80, log2FoldChange > logfc_threshold & padj <0.05)$gene,
                        'Hpr1' = filter(res_Hpr1, log2FoldChange > logfc_threshold & padj <0.05)$gene))
down_nsb2 <- fromList2(list("Cbp80"  = filter(res_Cbp80, log2FoldChange < -logfc_threshold& padj <0.05)$gene,
                            'Hpr1' = filter(res_Hpr1, log2FoldChange < -logfc_threshold& padj <0.05)$gene))


#identify genes in each group
Cbp80_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "Cbp80"))
Cbp80Hpr1_nsb_up <- format_intersect_members(get_intersect_members(up_nsb2, "Cbp80", "Hpr1"))
Cbp80_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "Cbp80"))
Cbp80Hpr1_nsb_down <- format_intersect_members(get_intersect_members(down_nsb2, "Cbp80", "Hpr1"))



# define set sizes for auto figure labelling by log2fc threshold
Cbp80_nsb_up_size <- nrow(Cbp80_nsb_up)
Cbp80Hpr1_nsb_up_size <- nrow(Cbp80Hpr1_nsb_up)
Cbp80_nsb_down_size <- nrow(Cbp80_nsb_down)
Cbp80Hpr1_nsb_down_size <- nrow(Cbp80Hpr1_nsb_down)

# write.xlsx(Cbp80_nsb_up, file="outputs/xlsx/upset_permutations_Cbp80_nsb_up.xlsx", sheetName="Cbp80_nsb_up", row.names=FALSE)
# write.xlsx(Cbp80Hpr1_nsb_up, file="outputs/xlsx/upset_permutations_Cbp80Hpr1_nsb_up.xlsx", sheetName="Cbp80Hpr1_nsb_up", append = T, row.names=FALSE)
# write.xlsx(Cbp80_nsb_down, file="outputs/xlsx/upset_permutations_Cbp80_nsb_down.xlsx", sheetName="Cbp80_nsb_down", row.names=FALSE)
# write.xlsx(Cbp80Hpr1_nsb_down, file="outputs/xlsx/upset_permutations_Cbp80Hpr1_nsb_down.xlsx", sheetName="Cbp80Hpr1_nsb_down", append = T, row.names=FALSE)

```








##FEATURES
Here we are interested in exploring the features that correspond to specific IPs (introns, Poly(A) length, ... etc)

```{r define_features for IP (Cbp80 and Cbp80/Hpr1 subsets)}
gtf <- import("inputs/GCF_000146045.2_R64_genomic.gtf")

features <- gtf %>%
  as.data.frame() %>%
  filter(type == "gene")

introns <- gtf %>% 
  as.data.frame() %>%
  filter(type == "exon") %>%
  group_by(gene_id) %>% 
  filter(n() > 1) %>% 
  ungroup() %>%
  dplyr::select(gene_id)
introns <- unique(introns$gene_id)

features <- features %>%
  mutate(introns = ifelse(gene_id %in% introns, "yes", "no"))

features <- features %>%
  dplyr::select(gene_id, width, strand, pseudo, introns)

norm_dds <- estimateSizeFactors(dds)
norm_counts <- DESeq2::counts(norm_dds, normalized=T)
polya_counts <- norm_counts %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  dplyr::select(gene, starts_with("2412_")) %>%
  rowwise() %>% 
  mutate(mean_polya = mean(c(`2412_INPUT_1_readcounts.txt`, `2412_INPUT_1_readcounts.txt`, `2412_INPUT_2_readcounts.txt`))) %>%
  dplyr::select(gene, mean_polya)

vsd <- vst(dds)
vsd <- assay(vsd)
polya_vsd <- vsd %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  dplyr::select(gene, starts_with("2412_")) %>%
  rowwise() %>% 
  mutate(mean_polya_vst = mean(c(`2412_INPUT_1_readcounts.txt`, `2412_INPUT_2_readcounts.txt`, `2412_INPUT_3_readcounts.txt`))) %>%
  dplyr::select(gene, mean_polya_vst)

features <- features %>%
  left_join(polya_counts, by = c("gene_id" = "gene")) %>%
  left_join(polya_vsd, by = c("gene_id" = "gene"))

# polyA
# 0.01 indicates transcripts that couldn't be measured bc there were fewer than 
# 5 transcripts in the library
polya_tails <- read_csv("https://data.mendeley.com/public-files/datasets/v5vm3dmm8y/files/aceb7ee8-5ba0-4cc5-be02-bdab3dc94cb3/file_downloaded") %>%
  dplyr::select(Systematic.Name, polya_length = WT.BY.mean)

features <- left_join(features, polya_tails, by = c("gene_id" = "Systematic.Name"))

# half life
rates <- read_tsv("https://rnajournal.cshlp.org/content/suppl/2014/08/08/rna.045104.114.DC1/TableS5.xls") %>%
  dplyr::select(gene_id = Syst, thalf, syn)
features <- left_join(features, rates, by = "gene_id")
```

```{r annotate_sets_with_feature}
# Results with INPUT background removed
# res_Cbp80 
# res_Cbp80Hpr1


res_Cbp80_features <- res_Cbp80 %>% filter(res_Cbp80$gene %in% Cbp80_nsb_up$gene | res_Cbp80$gene %in% Cbp80_nsb_down$gene) %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))
res_Cbp80_features$IP <- "Cbp80"

res_Cbp80Hpr1_features <- res_Hpr1 %>% filter(res_Hpr1$gene %in% Cbp80Hpr1_nsb_up$gene | res_Hpr1$gene %in% Cbp80Hpr1_nsb_down$gene) %>%
  left_join(features, by = c("gene" = "gene_id")) %>%
  mutate(change = ifelse(log2FoldChange > 0 , "log2fc_pos", "log2fc_neg"))
res_Cbp80Hpr1_features$IP <- "Cbp80-Hpr1"
  


```

```{r make_res_combined}
res_combined_features <- bind_rows(res_Cbp80_features, res_Cbp80Hpr1_features) %>%
  dplyr::select(gene, common, description,log2FoldChange, padj, width, strand, pseudo, introns, mean_polya, 
         mean_polya_vst, polya_length, thalf, syn, change, IP) %>%
  distinct()
res_PositiveCombined_features <- subset(res_combined_features, res_combined_features$change == "log2fc_pos")
```

## Length of transcripts

**Question**: 
Does transcript length impact transcript association with IP samples?

**Answer**: 
yes, genes that are enriched in Cbp80-Hpr1 pull downs are of longer length than those enriched in Cbp80-only pull downs

**Test used**:
wilcox.test() -- wilcoxon rank sum test. Is mean gene length different 
                 between up and down regulated genes? (data are not normally distributed)

```{r LENGTH_full_sets, fig.height=8, fig.width = 12}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different 
#                  between up and down regulated genes? (data are not normally distributed)

Cbp80_wilcox_length <- wilcox.test(res_Cbp80_features$width ~ res_Cbp80_features$change) %>%
  broom::tidy() %>%
  mutate(set = "Cbp80")

Cbp80Hpr1_wilcox_length <- wilcox.test(res_Cbp80Hpr1_features$width ~ res_Cbp80Hpr1_features$change) %>%
  broom::tidy() %>%
  mutate(set = "Hpr1")


UpCompare_wilcox_length <- wilcox.test(res_PositiveCombined_features$width ~ res_PositiveCombined_features$IP) %>%
  broom::tidy() %>%
  mutate(set = "Cbp80 v Cbp80-Hpr1")


all_wilcox_length <- rbind(Cbp80_wilcox_length, 
                           Cbp80Hpr1_wilcox_length,
                           UpCompare_wilcox_length) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)
  
plt1_length <- ggplot(res_Cbp80_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  ggtitle("Cbp80") + 
  xlim(0, 10000) +
  ylim(0, 0.0015)

plt2_length <- ggplot(res_Cbp80Hpr1_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  ggtitle("Cbp80-Hpr1")+ 
  xlim(0, 10000) +
  ylim(0, 0.0015)

plt3_length <- ggplot(res_PositiveCombined_features, aes(x = width)) +
  geom_density(alpha = .5, aes(fill = IP)) +
  geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  ggtitle("Cbp80-Hpr1")+ 
  xlim(0, 10000) +
  ylim(0, 0.001) +
  scale_fill_manual( values = c("Dodgerblue","darkorange"))

plt4_length <- ggtexttable(all_wilcox_length, rows = NULL, theme = ttheme(base_size = 7))

plt5_length <- ggplot(features, aes(x = width)) +
  theme_minimal() +
  geom_density() +
  xlim(0, 10000) +
  ylim(0, 0.0015)
  labs(title = "transcriptome distribution", x = "transcript length")

ggarrange1 <- ggarrange(plt1_length, plt2_length, 
                        nrow = 2, legend = "top", common.legend = T)

ggarrange2 <- ggarrange(plt4_length, plt5_length,
                        common.legend = T, nrow = 2, legend = "bottom")

ggarrange(ggarrange1, plt3_length, ggarrange2, ncol = 3, common.legend = T)

```


```{r LENGTH_by_shared_set2, fig.height=8, fig.width = 12}

plt1_length2 <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)), 
              aes(x = width, y = log2FoldChange, color = introns)) +
  geom_point(alpha = .5) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ", 
                 Cbp80_nsb_up_size, " + ",
                 Cbp80_nsb_down_size, ")")) +
  labs(x = "length") +
  scale_color_brewer(palette = "Set1")

plt2_length2 <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, 
                                   Cbp80Hpr1_nsb_down$gene)), 
              aes(x = width, y = log2FoldChange, color = introns)) +
  geom_point(alpha = .5) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80-Hpr1 (n = ",
          Cbp80Hpr1_nsb_up_size, " + ",
          Cbp80Hpr1_nsb_down_size, ")")) +
  labs(x = "length") +
  scale_color_brewer(palette = "Set1")



ggarrange(plt1_length2, plt2_length2, common.legend = T, legend = "bottom")
```



\newpage
## Introns 

**Question**: 
Are transcripts with introns associated with certain RBPs?

**Answer**: 
There is no signficant difference in the intron containing genes that associate with Cbp80-only or Cbp80-Hpr1

**Test Used**:
chisq.test() -- chi squared test. Are genes with introns 
                more or less likely to be associated with a different RBPs?

```{r INTRONS_full_sets, fig.height=8, fig.width = 12}


plt1_introns <- ggplot(res_Cbp80_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("Cbp80") +
  scale_fill_brewer(palette = "Set1")

plt2_introns <- ggplot(res_Cbp80Hpr1_features) +
  aes(x = change, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("Cbp80-Hpr1")+
  scale_fill_brewer(palette = "Set1")

plt3_introns <- ggplot(res_PositiveCombined_features) +
  aes(x = IP, fill = introns) +
  geom_bar(stat = "count") +
  geom_text(aes(label = ..count..), stat = "count") +
  theme_minimal() +
  ggtitle("Cbp80 v Cbp80-Hpr1") +
  scale_fill_manual( values = c("Dodgerblue","darkorange"))




ggarrange1 <- ggarrange(plt1_introns, plt2_introns, nrow = 1, legend = "bottom", common.legend = T)
ggarrange2 <- ggarrange(plt3_introns, common.legend = T, nrow =1, legend = "bottom")

ggarrange( ggarrange1, ggarrange2, ncol = 2, common.legend = T)
```

**Question**: What is the length of the intron-containing genes for genes enriched in the IP samples? are 

**Answer**: Of the genes that contain introns, there is no significant difference in length between those that associate with Cbp80-only and those that associate with Cbp80-Hpr1

```{r INTRON_lengths, fig.height=8, fig.width = 12}


res_PositiveCombined_Intron_features <- filter(res_PositiveCombined_features, introns == "yes")
UpCompare_intron_chisq <-  chisq.test(table(res_PositiveCombined_Intron_features$width, res_PositiveCombined_Intron_features$IP)) %>%
  tidy() %>%
  mutate(set = "Cbp80 v Cbp80-Hpr1")

all_intron_chisq <- UpCompare_intron_chisq %>%
  as.data.frame() %>%
  dplyr::select(-parameter) %>%
  mutate(method = "chisq.test()")



plt1_intron_len <- ggplot(res_Cbp80_features %>%
                            filter(log2FoldChange > 1 ), aes(x = width, fill = introns)) +
  geom_density(alpha = .5) +
  scale_fill_brewer(palette = "Set1") +
  #geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  xlim(c(0,7500)) +
  ylim(c(0,0.001)) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(x = "length", title = "Cbp80 Only, log2FC>1")

plt2_intron_len <- ggplot(res_Cbp80Hpr1_features %>%
                            filter(log2FoldChange > 1 ), aes(x = width, fill = introns)) +
  geom_density(alpha = .5) +
  scale_fill_brewer(palette = "Set1") +
  #geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  xlim(c(0,7500)) +
  ylim(c(0,0.001)) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(x = "length", title = "Cbp80-Hpr1, log2FC>1")

plt3_intron_len <- ggplot(res_PositiveCombined_Intron_features %>%
                            filter(log2FoldChange > 1 ), aes(x = width, fill = IP)) +
  geom_density(alpha = .5) +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  #geom_density(data = features, aes(x = width), linetype = "dashed", color ="black") +
  theme_minimal() +
  xlim(c(0,7500)) +
  ylim(c(0,0.001)) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  labs(x = "length", title = "Cbp80 v Cbp80-Hpr1, +introns, log2FC>1")

plt4_introns <- ggtexttable(all_intron_chisq, rows = NULL, theme = ttheme(base_size = 8))



ggarrange1 <- ggarrange(plt1_intron_len, plt2_intron_len, nrow = 2, common.legend = T, legend = "bottom")
ggarrange2 <- ggarrange(plt4_introns, plt3_intron_len, nrow = 2, common.legend = T, legend = "right")

ggarrange( ggarrange1, ggarrange2, ncol = 2)
```


```{r INTRONS_LENGTH_regression_model, fig.height=8, fig.width = 12}
res_Cbp80_features$introns2 <- ifelse(res_Cbp80_features$introns == "yes", 1, 0)
Cbp80_width_intron <- lm(formula = log2FoldChange ~ width + introns2, data = res_Cbp80_features)
Cbp80_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_Cbp80_features %>% filter(polya_length > 1), na.action = "na.omit")
Cbp80_width_intron %>% glance()
Cbp80_width_intron %>% tidy()

res_Cbp80Hpr1_features$introns2 <- ifelse(res_Cbp80Hpr1_features$introns == "yes", 1, 0)
Cbp80Hpr1_width_intron <- lm(formula = log2FoldChange ~ width + introns2 + polya_length + mean_polya_vst, data = res_Cbp80Hpr1_features %>% filter(polya_length > 1), na.action = "na.omit")
Cbp80Hpr1_width_intron %>% glance()
Cbp80Hpr1_width_intron %>% tidy()
```

\newpage
## Poly(A) tail length

**Question**:
Do the transcripts that associate with a particular IP have longer/shorter tail lengths

**Answer**:
Poly(A) tails are not significantly longer/shorter for transcripts in either IP

**Test used**:
wilcox.test() -- wilcoxon rank sum test. Is mean poly(A) tail length different
                 between two groups? (data are not normally distributed)

```{r POLYA_by_shared_set, fig.height=8, fig.width = 12}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

Cbp80_wilcox_polya2 <- res_Cbp80_features %>%
  filter(gene %in% c(Cbp80_nsb_up$gene, 
                     Cbp80_nsb_down$gene)) %>%
  filter(polya_length > 1) %>%
  do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
  mutate(set = "Cbp80")

Cbp80Hpr1_wilcox_polya2 <- res_Cbp80Hpr1_features %>%
  filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)) %>%
  filter(polya_length > 1) %>%
  do(tidy(wilcox.test(polya_length ~ change, data = .))) %>%
  mutate(set = "Cbp80Hpr1")

UpCompare_wilcox_polya2 <- res_PositiveCombined_features %>%
  filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80_nsb_up$gene)) %>%
  filter(polya_length > 1) %>%
  do(tidy(wilcox.test(polya_length ~ IP, data = .))) %>%
  mutate(set = "Cbp80 v Cbp80Hpr1")


all_wilcox_polya2 <- rbind(Cbp80_wilcox_polya2,
                           Cbp80Hpr1_wilcox_polya2,
                           UpCompare_wilcox_polya2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)



plt1_polya <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ",
          Cbp80_nsb_up_size, " + ",
          Cbp80_nsb_down_size, ")"))

plt2_polya <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80Hpr1 (n = ",
                 Cbp80Hpr1_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_down_size, ")")) 


plt3_polya <- ggplot(res_PositiveCombined_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length)) +
  geom_density(alpha = .5, aes(fill = IP)) +
  geom_density(data = features %>% filter(polya_length > 1), aes(x = polya_length), linetype = "dashed", color = "black") +
  theme_minimal() +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 v Cbp80Hpr1 (n = ",
                 Cbp80_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_up_size, ")")) 


plt4_polya <- ggtexttable(all_wilcox_polya2, rows = NULL, theme = ttheme(base_size = 8))

# plt5_polya <- ggplot(features %>%
#                        filter(polya_length > 1), 
#                      aes(x = polya_length)) +
#   theme_minimal() +
#   geom_density() +
#   labs(title = "transcriptome distribution", x = "poly(A) tail length")




ggarrange1 <- ggarrange(plt1_polya, plt2_polya, nrow = 2, common.legend = T, legend = "bottom")

ggarrange2 <- ggarrange(plt4_polya, plt3_polya, nrow = 2, common.legend = T, legend = "bottom")

ggarrange( ggarrange1, ggarrange2, ncol = 3)


```

```{r POLYA_2, fig.height=8, fig.width = 12}
plt5_polya <- ggplot(features %>%
                      filter(polya_length > 1), 
                    aes(y = width, x = polya_length)) +
  geom_point(alpha = .1) +
  theme_minimal() +
  labs(y = "transcript length", x = "poly(A) tail length", title = "whole transcriptome")


plt6_polya <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)) %>%
                        filter(polya_length > 1), 
             aes(x = polya_length, y = width)) +
  geom_point(alpha = .5, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ",
          Cbp80_nsb_up_size, " + ",
          Cbp80_nsb_down_size, ")")) +
  labs(y = "transcript length", x = "poly(A) tail length")

plt7_polya <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length, y = width)) +
  geom_point(alpha = .5, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80Hpr1 (n = ",
                 Cbp80Hpr1_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_down_size, ")"))  +
  labs(y = "transcript length", x = "poly(A) tail length")

plt8_polya <- ggplot(res_PositiveCombined_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene))%>%
                        filter(polya_length > 1), 
              aes(x = polya_length, y = width)) +
  geom_point(alpha = .6, aes(color = IP)) +
  theme_minimal() +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 v Cbp80Hpr1 (n = ",
                 Cbp80_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_up_size, ")"))  +
  labs(y = "transcript length", x = "poly(A) tail length")

ggarrange1 <- ggarrange(plt6_polya, plt7_polya, nrow = 2 ,
          common.legend = T, legend = "bottom")

ggarrange2 <- ggarrange(plt5_polya, plt8_polya, nrow = 2 ,
          common.legend = T, legend = "bottom")

ggarrange(ggarrange1, ggarrange2)
```
\newpage
## Synthesis rate

"Note because the x axis is log10 transformed, synthesis time = 0 are NOT on the density plots.
I performed the statistics on non-log10 transformed synthesis rates." - Taylor R

**Question**:
Are the synthesis rates unique to different RBPs?

**Answer**:
transcripts associated with Cbp80-Hpr1 have a signficantly higher synthesis rate


**Tested Used**:
wilcox.test() -- wilcoxon rank sum test. Is the mean synthesis rate different
                 between two groups? (data are not normally distributed)

```{r SYNTHESIS_by_shared_set, fig.height=8, fig.width = 12}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

Cbp80_wilcox_synth2 <- res_Cbp80_features %>%
  filter(gene %in% c(Cbp80_nsb_up$gene, 
                     Cbp80_nsb_down$gene)) %>%
  do(tidy(wilcox.test(syn ~ change, data = .))) %>%
  mutate(set = "Cbp80")

Cbp80Hpr1_wilcox_synth2 <- res_Cbp80Hpr1_features %>%
  filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)) %>%
  do(tidy(wilcox.test(syn ~ change, data = .))) %>%
  mutate(set = "Cbp80Hpr1")

UpCompare_wilcox_synth2 <- res_PositiveCombined_features %>%
  filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)) %>%
  do(tidy(wilcox.test(syn ~ IP, data = .))) %>%
  mutate(set = "Cbp80 v Cbp80Hpr1")

all_wilcox_synth2 <- rbind(Cbp80_wilcox_synth2,
                           Cbp80Hpr1_wilcox_synth2,
                           UpCompare_wilcox_synth2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)


plt1_synth <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)), 
              aes(x = syn)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = syn), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ",
          Cbp80_nsb_up_size, " + ",
          Cbp80_nsb_down_size, ")"))+
  labs(x = "log10 synthesis rate") +
  scale_x_log10()

plt2_synth <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)), 
              aes(x = syn)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = syn), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80Hpr1 (n = ",
                 Cbp80Hpr1_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_down_size, ")")) +
  labs(x = "log10 synthesis rate") +
  scale_x_log10()

plt3_synth <- ggplot(res_PositiveCombined_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)), 
              aes(x = syn)) +
  geom_density(alpha = .5, aes(fill = IP)) +
  geom_density(data = features, aes(x = syn), linetype = "dashed", color = "black") +
  theme_minimal() +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 v Cbp80Hpr1 (n = ",
                 Cbp80_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_up_size, ")")) +
  labs(x = "log10 synthesis rate") +
  scale_x_log10()


plt4_synth <- ggtexttable(all_wilcox_synth2, rows = NULL, theme = ttheme(base_size = 8))

# plt5_synth <- ggplot(features, aes(x = syn)) +
#   theme_minimal() +
#   geom_density() +
#   labs(title = "transcriptome distribution", x = "log10 synthesis rate") +
#   scale_x_log10()



ggarrange1 <- ggarrange(plt1_synth, plt2_synth,
                        legend = "bottom", common.legend = T, nrow = 2)


ggarrange2 <- ggarrange(plt4_synth, plt3_synth,
                        legend = "bottom", common.legend = T, nrow = 2)


ggarrange(ggarrange1, ggarrange2, ncol = 2)
```

```{r SYNTHESIS_2, fig.height=8, fig.width = 12}
plt6_synth <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)), 
             aes(x = syn, y = log2FoldChange)) +
  geom_point(alpha = .4, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ",
          Cbp80_nsb_up_size, " + ",
          Cbp80_nsb_down_size, ")")) +
  labs(y = "log2 Fold Change", x = "log10 synthesis rate") +
  scale_x_log10()

plt7_synth <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)), 
              aes(x = syn, y = log2FoldChange)) +
  geom_point(alpha = .4, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80Hpr1 (n = ",
                 Cbp80Hpr1_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_down_size, ")")) +
  labs(y = "log2 Fold Change", x = "log10 synthesis rate") +
  scale_x_log10()

plt8_synth <- ggplot(res_PositiveCombined_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)), 
              aes(x = syn, y = log2FoldChange)) +
  geom_point(alpha = .4, aes(color = IP)) +
  theme_minimal() +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 v Cbp80Hpr1 (n = ",
                 Cbp80_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_up_size, ")")) +
  labs(y = "log2 Fold Change", x = "log10 synthesis rate") +
  scale_x_log10()


ggarrange1 <- ggarrange(plt6_synth, plt7_synth, nrow = 2,
          common.legend = T, legend = "bottom")
ggarrange2 <- ggarrange(plt8_synth,
          common.legend = T, legend = "bottom")
 
ggarrange(ggarrange1, ggarrange2, ncol = 2,
          common.legend = T, legend = "bottom")
```
\newpage
## Half life

***Question***:
Are transcripts with longer/shorter half-life more likely to be associated with specific RBP subsets. 

***Answer***:
Cbp80-only associate transcripts have a significantly longer half-life

***Test Used***:
wilcox.test() -- wilcoxon rank sum test. Is the mean half-life different
                 between two groups? (data are not normally distributed)

```{r HALFLIFE_by_shared_set, fig.height=8, fig.width = 12}
# wilcox.test() -- wilcoxon rank sum test. Is mean gene length different
#                  between two groups? (data are not normally distributed)

Cbp80_wilcox_thalf2 <- res_Cbp80_features %>%
  filter(gene %in% c(Cbp80_nsb_up$gene, 
                     Cbp80_nsb_down$gene)) %>%
  do(tidy(wilcox.test(thalf ~ change, data = .))) %>%
  mutate(set = "Cbp80")

Cbp80Hpr1_wilcox_thalf2 <- res_Cbp80Hpr1_features %>%
  filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)) %>%
  do(tidy(wilcox.test(thalf ~ change, data = .))) %>%
  mutate(set = "Cbp80-Hpr1")

UpCompare_wilcox_thalf2 <- res_PositiveCombined_features %>%
  filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)) %>%
  do(tidy(wilcox.test(thalf ~ IP, data = .))) %>%
  mutate(set = "Cbp80 v Cbp80-Hpr1")

all_wilcox_thalf2 <- rbind(Cbp80_wilcox_thalf2,
                           Cbp80Hpr1_wilcox_thalf2,
                           UpCompare_wilcox_thalf2) %>%
  as.data.frame() %>%
  mutate(method = "wilcox.test()") %>%
  dplyr::select(statistic, p.value, method, set)

plt1_thalf <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)), 
              aes(x = thalf)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = thalf), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ",
          Cbp80_nsb_up_size, " + ",
          Cbp80_nsb_down_size, ")")) +
  labs(x = "thalf") 

plt2_thalf <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)), 
              aes(x = thalf)) +
  geom_density(alpha = .5, aes(fill = change)) +
  geom_density(data = features, aes(x = thalf), linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80Hpr1 (n = ",
                 Cbp80Hpr1_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_down_size, ")")) +
  labs(x = "thalf")

plt3_thalf <- ggplot(res_PositiveCombined_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)), 
              aes(x = thalf)) +
  geom_density(alpha = .5, aes(fill = IP)) +
  geom_density(data = features, aes(x = thalf), linetype = "dashed", color = "black") +
  theme_minimal() +
  xlim(c(0,50)) +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 v Cbp80Hpr1 (n = ",
                 Cbp80_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_up_size, ")")) +
  labs(x = "thalf")


plt4_thalf <- ggtexttable(all_wilcox_thalf2, rows = NULL, theme = ttheme(base_size = 8))

ggarrange1 <- ggarrange(plt1_thalf,  plt2_thalf, nrow = 2, 
          common.legend = T, legend = "bottom")

ggarrange2 <- ggarrange(plt4_thalf,  plt3_thalf, nrow = 2, 
          common.legend = T, legend = "bottom")

ggarrange(ggarrange1, ggarrange2, ncol = 2,
          common.legend = T, legend = "bottom")
```

```{r HALFLIFE_2, fig.height=8, fig.width = 12}
plt5_thalf <- ggplot(res_Cbp80_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, 
                                   Cbp80_nsb_down$gene)), 
             aes(x = thalf, y = log2FoldChange)) +
  geom_point(alpha = .3, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 (n = ",
          Cbp80_nsb_up_size, " + ",
          Cbp80_nsb_down_size, ")")) 

plt6_thalf <- ggplot(res_Cbp80Hpr1_features %>%
                filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)), 
              aes(x = thalf, y = log2FoldChange)) +
  geom_point(alpha = .3, aes(color = change)) +
  theme_minimal() +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80Hpr1 (n = ",
                 Cbp80Hpr1_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_down_size, ")"))

plt7_thalf <- ggplot(res_PositiveCombined_features %>%
                filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)), 
              aes(x = thalf, y = log2FoldChange)) +
  geom_point(alpha = .3, aes(color = IP)) +
  theme_minimal() +
  xlim(c(0,50)) +
  scale_fill_manual( values = c("Dodgerblue","darkorange")) +
  theme(plot.title = element_text(size=9),
        plot.title.position = "plot") +
  ggtitle(paste0("Cbp80 v Cbp80Hpr1 (n = ",
                 Cbp80_nsb_up_size, " + ", 
                 Cbp80Hpr1_nsb_up_size, ")"))
 

ggarrange1 <- ggarrange(plt5_thalf, plt6_thalf, nrow = 2,
          common.legend = T, legend = "bottom")

ggarrange2 <- ggarrange(plt7_thalf, 
          common.legend = T, legend = "bottom")

ggarrange(ggarrange1, ggarrange2)

```
\newpage
## Do all of these features correlate?

(E.g. are we pulling the same pattern over and over?)

Looks like most features overlap between Cbp80 and Cbp80-Hpr1


```{r CORRELATE, fig.height=8, fig.width = 12}
library(GGally)

# cor_Cbp80_features <- res_Cbp80_features %>%
#   filter(gene %in% c(Cbp80_nsb_up$gene, 
#                      Cbp80_nsb_down$gene)) %>%
#   dplyr::select(change, transcript_length = width, polya_length, log2FoldChange,
#          syn, thalf, mean_polya_vst) 
# 
# ggpairs(cor_Cbp80_features, columns = 2:7,
#        ggplot2::aes(colour=change, alpha = .25), showStrips = T,
#        title = paste0("Cbp80 (n = ",
#                       Cbp80_nsb_up_size, " + ",
#                       Cbp80_nsb_down_size, ")")) +
#   theme_minimal()
# 
# cor_Cbp80Hpr1_features <- res_Cbp80Hpr1_features %>%
#   filter(gene %in% c(Cbp80Hpr1_nsb_up$gene, Cbp80Hpr1_nsb_down$gene)) %>%
#   dplyr::select(change, transcript_length = width, polya_length, log2FoldChange,
#          syn, thalf, mean_polya_vst) 
# 
# ggpairs(cor_Cbp80Hpr1_features, columns = 2:7,
#        ggplot2::aes(colour=change, alpha = .25), showStrips = T,
#        title = paste0("Cbp80Hpr1 (n = ",
#                       Cbp80Hpr1_nsb_up_size, " + ", 
#                       Cbp80Hpr1_nsb_down_size, ")")) +
#   theme_minimal()

cor_UpCompare_features <- res_PositiveCombined_features %>%
  filter(gene %in% c(Cbp80_nsb_up$gene, Cbp80Hpr1_nsb_up$gene)) %>%
  dplyr::select(change, transcript_length = width, polya_length, log2FoldChange,
         syn, thalf, mean_polya_vst, IP) 

ggpairs(cor_UpCompare_features, columns = 2:7,
       ggplot2::aes(colour=IP, alpha = .25), showStrips = T,
       title = paste0("Cbp80 v Cbp80Hpr1 (n = ",
                      Cbp80_nsb_up_size, " + ", 
                      Cbp80Hpr1_nsb_up_size, ")")) +
  theme_minimal()
```
\newpage
## Enrichment

**Question**: 
What are the GO terms enriched with each RBPs subset?


```{r enrichment_functions}
res_enrichment <- function(res, title){
  # separate to induced and repressed
  up <- res[res$log2FoldChange > 2, ]
  down <- res[res$log2FoldChange < -2, ] 
  
  
  # GO enrichment
  ego_up <- enrichGO(up$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                     ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_up@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt3 <- dotplot(ego_up, showCategory = plt_rows, font.size = 5) + theme_minimal(base_size = 5)
  
  ego_down <- enrichGO(down$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                       ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_down@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt4 <- dotplot(ego_down, showCategory = plt_rows, font.size = 5) + theme_minimal(base_size = 5)
  
  # combine plots
  all_plts <- ggarrange(plt3, plt4, ncol = 2, nrow = 1, common.legend = F, 
                       legend = "bottom", labels = c("A: log2FC +", "B: log2FC -"),
                       heights = c(1, 2))
  annotate_figure(all_plts, top = text_grob(title))
}

enrichment_intersect <- function(up, down, title){
  # GO enrichment
  require(org.Sc.sgd.db)
  ego_up <- enrichGO(up$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                     ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_up@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt1 <- dotplot(ego_up, showCategory = plt_rows, font.size = 6, split="ONTOLOGY") + 
    facet_grid(ONTOLOGY~., scale="free") +
    theme_minimal(base_size = 6)
  
  ego_down <- enrichGO(down$common, OrgDb = org.Sc.sgd.db, keyType = "GENENAME", 
                       ont = "ALL", pAdjustMethod = "bonferroni")
  plt_rows <- ego_down@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt2 <- dotplot(ego_down, showCategory = plt_rows, font.size = 6, split="ONTOLOGY") + 
    facet_grid(ONTOLOGY~., scale="free") +
    theme_minimal(base_size = 6)
  # combine plots
  all_plts<- ggarrange(plt1, plt2, ncol = 2, nrow = 1, common.legend = F, 
                       legend = "bottom", labels = c("A: log2FC +", "B: log2FC -"),
                       heights = c(1, 2))
  annotate_figure(all_plts, top = text_grob(title))
}

rna_classes <- read_csv("https://osf.io/gqnwh/download", skip = 1) %>%
  dplyr::select("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X") %>%
  pivot_longer(cols = I:X, names_to = "rna_class", values_to = "gene") %>%
  filter(!grepl("CUT", gene)) %>%
  filter(!grepl("SUT", gene)) 

enrichment_intersect_rna_classes <- function(up, down, title){
  enricher_up <- enricher(gene = up$common, TERM2GENE = rna_classes, maxGSSize = 703)
  plt_rows <- enricher_up@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt1 <- dotplot(enricher_up, showCategory = plt_rows, font.size = 8) + 
    theme_minimal(base_size = 6)
  
  enricher_down <- enricher(gene = down$common, TERM2GENE = rna_classes, maxGSSize = 703)
  plt_rows <- enricher_down@result %>%
    filter(p.adjust < .05) %>%
    nrow()
  plt2 <- dotplot(enricher_down, showCategory = plt_rows, font.size = 8) + 
    theme_minimal(base_size = 6)
  # combine plots
  all_plts<- ggarrange(plt1, plt2, ncol = 2, nrow = 1, common.legend = F, 
                       legend = "bottom", labels = c("A: log2FC +", "B: log2FC -"),
                       heights = c(1, 2))
  annotate_figure(all_plts, top = text_grob(title))
}
```

### full sets

```{r GO_ENRICHMENT_full_sets, fig.height=8, fig.width = 12}
enrichment_intersect(up = filter(res_Cbp80, log2FoldChange > 1 & padj < 0.05),
                    down = filter(res_Cbp80, log2FoldChange < 1 & padj < 0.05),
                    title = "Cbp80")

```
\newpage

```{r GO_ENRICHMENT_full_sets2, fig.height=8, fig.width = 12}
res_Cbp80Hpr1 <- filter(res_Hpr1, gene %in% res_Cbp80Hpr1_features$gene)
enrichment_intersect(up = filter(res_Cbp80Hpr1, log2FoldChange > 1 & padj < 0.05),
                     down = filter(res_Cbp80Hpr1, log2FoldChange < 1 & padj < 0.05),
                     title = "Cbp80-Hpr1")


```

```{r RNA_CLASSES_full_sets, fig.height=8, fig.width = 12}
enrichment_intersect_rna_classes(up = filter(res_Cbp80, log2FoldChange > 1 & padj < 0.05),
                                 down = filter(res_Cbp80, log2FoldChange < 1 & padj < 0.05),
                                 title = "Cbp80")

enrichment_intersect_rna_classes(up = filter(res_Cbp80Hpr1, log2FoldChange > 1 & padj < 0.05),
                                 down = filter(res_Cbp80Hpr1, log2FoldChange < 1 & padj < 0.05),
                                 title = "Cbp80-Hpr1")
```
